### PR TITLE
[SYCLomatic] Support migration for 24 cuSPARSE API 

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei zhiwei.jiang@intel.com

### CODE REVIEW

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 1]
```diff
@@ -294,6 +294,75 @@ class optimize_info {
...
+    auto data_row_ptr = dpct::detail::get_memory<int>(row_ptr);
+    auto data_col_ind = dpct::detail::get_memory<int>(col_ind);
+    auto data_val = dpct::detail::get_memory<Ty>(val);
+    oneapi::mkl::sparse::set_csr_data(queue, optimize_info->get_matrix_handle(),
+                                      row_col, row_col, info->get_index_base(),
+                                      data_row_ptr, data_col_ind, data_val);
+    if (info->get_matrix_type() != matrix_info::matrix_type::tr)
+      return;
+#ifndef DPCT_USM_LEVEL_NONE
+    sycl::event e;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: can use macro SPARSE_CALL

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 2]
```diff
@@ -294,6 +294,75 @@ class optimize_info {
...
+                  void *y) {
+    using Ty = typename dpct::DataType<T>::T2;
+    if (info->get_matrix_type() != matrix_info::matrix_type::tr)
+      return;
+    auto alpha_value =
+        dpct::detail::get_value(static_cast<const Ty *>(alpha), queue);
+    Ty *new_x_ptr = nullptr;
+    if (alpha_value != Ty(1.0f)) {
+      new_x_ptr = (Ty *)dpct::dpct_malloc(row_col * sizeof(Ty));
+      dpct::dpct_memcpy(new_x_ptr, x, row_col * sizeof(Ty));
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: can use async version

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 3]
```diff
@@ -294,6 +294,75 @@ class optimize_info {
...
+    e =
+#endif
+        oneapi::mkl::sparse::trsv(
+            queue, info->get_uplo(), trans, info->get_diag(),
+            optimize_info->get_matrix_handle(), data_new_x, data_y);
+#ifndef DPCT_USM_LEVEL_NONE
+    optimize_info->add_dependency(e);
+#endif
+    if (alpha_value != Ty(1.0f)) {
+      queue.wait();
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: can use async version

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 4]
```diff
@@ -294,6 +294,75 @@ class optimize_info {
...
+                  std::shared_ptr<optimize_info> optimize_info) {
+    using Ty = typename dpct::DataType<T>::T2;
+    auto data_row_ptr = dpct::detail::get_memory<int>(row_ptr);
+    auto data_col_ind = dpct::detail::get_memory<int>(col_ind);
+    auto data_val = dpct::detail::get_memory<Ty>(val);
+    oneapi::mkl::sparse::set_csr_data(queue, optimize_info->get_matrix_handle(),
+                                      row_col, row_col, info->get_index_base(),
+                                      data_row_ptr, data_col_ind, data_val);
+    if (info->get_matrix_type() != matrix_info::matrix_type::tr)
+      return;
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: throw exception

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -3693,13 +3693,25 @@ void SPBLASFunctionCallRule::registerMatcher(MatchFinder &MF) {
...
-        "cusparseDestroySolveAnalysisInfo",
+        "cusparseDestroySolveAnalysisInfo", "cusparseCreateCsrsv2Info",
+        "cusparseDestroyCsrsv2Info",
         /*level 2*/
         "cusparseScsrmv", "cusparseDcsrmv", "cusparseCcsrmv", "cusparseZcsrmv",
         "cusparseScsrmv_mp", "cusparseDcsrmv_mp", "cusparseCcsrmv_mp",
         "cusparseZcsrmv_mp", "cusparseCsrmvEx", "cusparseCsrmvEx_bufferSize",
         "cusparseScsrsv_analysis", "cusparseDcsrsv_analysis",
         "cusparseCcsrsv_analysis", "cusparseZcsrsv_analysis",
+        "cusparseScsrsv_solve", "cusparseDcsrsv_solve", "cusparseCcsrsv_solve",
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:3704]: Please consider refactor the code by put all cusparse* migration logic into single files.

File: clang/lib/DPCT/LIBCUAPIMigration.cpp [hunk 6]
```diff
@@ -101,5 +101,26 @@ void LIBCURule::runRule(const ast_matchers::MatchFinder::MatchResult &Result) {
   EA.applyAllSubExprRepl();
 }
 
+void LibraryTypeLocRule::registerMatcher(ast_matchers::MatchFinder &MF) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/LIBCUAPIMigration.cpp]: please put the code into LibraryAPIMigration.cpp, or create a new.

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 7]
```diff
@@ -3602,7 +3602,7 @@ REGISTER_RULE(RandomEnumsRule, PassKind::PK_Migration, RuleGroupKind::RK_Rng)
 void SPBLASEnumsRule::registerMatcher(MatchFinder &MF) {
   MF.addMatcher(declRefExpr(to(enumConstantDecl(matchesName(
                                 "(CUSPARSE_STATUS.*)|(CUSPARSE_POINTER_MODE.*)|"
-                                "(CUSPARSE_ALG.*)"))))
+                                "(CUSPARSE_ALG.*)|(CUSPARSE_SOLVE_POLICY.*)"))))
```
Comment (zhimingwang36) [inline clang/lib/DPCT/ASTTraversal.cpp:3605]: Please consider a follow up action to refactor the code, put migration rules for same lib migration into same single file.

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 8]
```diff
@@ -294,6 +294,81 @@ class optimize_info {
 #endif
 
 #ifdef __INTEL_MKL__ // The oneMKL Interfaces Project does not support this.
+namespace detail {
+#ifdef DPCT_USM_LEVEL_NONE
+#define SPARSE_CALL(X, Y) X;
+#else
+#define SPARSE_CALL(X, Y)                                                      \
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]:
```suggestion
#define SPARSE_CALL(Call, Handle)                                                      \
```

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 9]
```diff
@@ -294,6 +294,81 @@ class optimize_info {
...
+  sycl::event e = X;                                                           \
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: Line 302 can be removed by inline X in the call.

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 7 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp:3605]: OK

File: clang/lib/DPCT/LIBCUAPIMigration.cpp (hunk 6 above)
Comment (zhiweij1) [reply clang/lib/DPCT/LIBCUAPIMigration.cpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 8 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp (hunk 9 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 10]
```diff
@@ -294,6 +294,79 @@ class optimize_info {
 #endif
 
 #ifdef __INTEL_MKL__ // The oneMKL Interfaces Project does not support this.
+namespace detail {
+#ifdef DPCT_USM_LEVEL_NONE
+#define SPARSE_CALL(CALL, OPTIMIZE_INFO) CALL;
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: name OPTIMIZE_INFO doens't make sense for all use points.  Need a common name.

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: done

File: clang/runtime/dpct-rt/include/sparse_utils.hpp [hunk 11]
```diff
@@ -294,6 +294,81 @@ class optimize_info {
...
+  void operator()(sycl::queue &queue, oneapi::mkl::transpose trans, int row_col,
+                  const void *alpha, const std::shared_ptr<matrix_info> info,
+                  const void *val, const int *row_ptr, const int *col_ind,
+                  std::shared_ptr<optimize_info> optimize_info, const void *x,
+                  void *y) {
+    using Ty = typename dpct::DataType<T>::T2;
+    if (info->get_matrix_type() != matrix_info::matrix_type::tr)
+      throw std::runtime_error(
+          "dpct::sparse::csrsv_impl()(): oneapi::mkl::sparse::trsv "
+          "only accept triangular matrix.");
```
Comment (spencerpatty) [inline clang/runtime/dpct-rt/include/sparse_utils.hpp]: in oneMKL, this is not necessary, we respect the uplo parameter to signal which part of matrix data to utilize.  If itis general, then we pick up the appropriate part based on what is actually there in the data representation.  We don't however support symmetric where the data is not there for whatever uplo value you are using ...

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/sparse_utils.hpp]: OK, I will remove it.

**Review by spencerpatty — APPROVED** [review]
csrsv changes look fine to me.  Approved

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**